### ISSUE TITLE
Custom class in SYCL device code makes the linker crash

### ISSUE THREAD

**Issue by krasznaa** [intel/llvm#1616]
I ran into an issue with the following simple example:

```c++
// SYCL include(s).
#include <CL/sycl.hpp>

// System include(s).
#include <cmath>
#include <iostream>
#include <vector>

/// Dummy class helping with multiplying floats
class MultiplierHelper {

public:
   /// Constructor
   MultiplierHelper( float multiplier )
#ifdef HELPER_IS_INLINE
   : m_multiplier( multiplier ) {}
#else
   ;
#endif
   /// Perform the multiplication
   float process( float value ) const
#ifdef HELPER_IS_INLINE
   { return value * m_multiplier; }
#else
   ;
#endif

private:
   float m_multiplier; ///< Multiplier to use

}; // class MultiplierHelper

#ifndef HELPER_IS_INLINE
MultiplierHelper::MultiplierHelper( float multiplier )
: m_multiplier( multiplier ) {

}

float MultiplierHelper::process( float value ) const {

   return value * m_multiplier;
}
#endif // not HELPER_IS_INLINE

int main() {

   // Create a vector array that would be manipulated.
   std::vector< float > dummyArray;
   static const std::size_t ARRAY_SIZE = 10000;
   dummyArray.reserve( ARRAY_SIZE );
   static const float ARRAY_ELEMENT = 3.141592f;
   for( std::size_t i = 0; i < ARRAY_SIZE; ++i ) {
      dummyArray.push_back( ARRAY_ELEMENT );
   }

   // Set up a SYCL buffer on top of this STL object.
   cl::sycl::buffer< cl::sycl::cl_float > buffer( dummyArray.begin(),
                                                  dummyArray.end() );

   // Set up the SYCL queue.
   cl::sycl::queue queue;
   cl::sycl::range< 1 > workItems( buffer.get_count() );

   // Let the user know what device the calculation is running on.
   const cl::sycl::device& device = queue.get_device();
   std::cout << "Using device: "
             << device.get_info< cl::sycl::info::device::name >()
             << " (" << device.get_info< cl::sycl::info::device::version >()
             << ")" << std::endl;

   // Set up the helper object.
   static const float MULTIPLIER = 1.2345f;

   // Run a job using the helper object.
   queue.submit( [&]( cl::sycl::handler& handler ) {
         auto acc =
            buffer.get_access< cl::sycl::access::mode::read_write >( handler );
         handler.parallel_for< class SYCLMultiply >( workItems,
            [=]( cl::sycl::id< 1 > id ) {
               MultiplierHelper helper( MULTIPLIER );
               acc[ id ] = helper.process( acc[ id ] );
            } );
      } );

   // Copy the results back to the host. Note that the execution will wait
   // here for a potentially offloaded operation to finish first. Put the
   // reader accessor into a scope, to make sure it is deleted after it is
   // not needed anymore.
   {
      auto bReader =
         buffer.get_access< cl::sycl::access::mode::read >();
      for( std::size_t i = 0; i < buffer.get_count(); ++i ) {
         dummyArray[ i ] = bReader[ i ];
      }
   }

   // Check those results.
   static const float EXPECTED_RESULT = ARRAY_ELEMENT * MULTIPLIER;
   for( std::size_t i = 0; i < ARRAY_SIZE; ++i ) {
      if( std::abs( dummyArray[ i ] - EXPECTED_RESULT ) > 0.001 ) {
         std::cerr << "The SYCL transformation failed to run" << std::endl;
         return 1;
      }
   }
   std::cout << "The SYCL transformation succeeded" << std::endl;

   // Return gracefully.
   return 0;
}
```

Trying to compile it with the latest version of the code, I get:

```
[bash][Elrond]:OutOfSource > clang++ -fsycl -o linker_error linker_error.cxx 
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.      Program arguments: /media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv -o /tmp/linker_error-a887e3-608d7b.spv -spirv-max-version=1.1 -spirv-ext=+all /tmp/linker_error-253ee5.bc 
1.      Running pass 'LLVMToSPIRV' on module '/tmp/linker_error-253ee5.bc'.
 #0 0x000055f94edffc1a llvm::sys::PrintStackTrace(llvm::raw_ostream&) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x51ac1a)
 #1 0x000055f94edfdba4 llvm::sys::RunSignalHandlers() (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x518ba4)
 #2 0x000055f94edfdce8 SignalHandler(int) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x518ce8)
 #3 0x00007faadba1d890 __restore_rt (/lib/x86_64-linux-gnu/libpthread.so.0+0x12890)
 #4 0x000055f94ed7fb70 llvm::Value::getName() const (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x49ab70)
 #5 0x000055f94ea77c1e SPIRV::LLVMToSPIRV::transDirectCallInst(llvm::CallInst*, SPIRV::SPIRVBasicBlock*) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x192c1e)
 #6 0x000055f94ea70514 SPIRV::LLVMToSPIRV::transValueWithoutDecoration(llvm::Value*, SPIRV::SPIRVBasicBlock*, bool) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x18b514)
 #7 0x000055f94ea70919 SPIRV::LLVMToSPIRV::transValue(llvm::Value*, SPIRV::SPIRVBasicBlock*, bool) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x18b919)
 #8 0x000055f94ea76771 SPIRV::LLVMToSPIRV::transFunction(llvm::Function*) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x191771)
 #9 0x000055f94ea76c04 SPIRV::LLVMToSPIRV::translate() (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x191c04)
#10 0x000055f94ea76dac SPIRV::LLVMToSPIRV::runOnModule(llvm::Module&) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x191dac)
#11 0x000055f94ed53a27 llvm::legacy::PassManagerImpl::run(llvm::Module&) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x46ea27)
#12 0x000055f94ea78152 llvm::writeSpirv(llvm::Module*, SPIRV::TranslatorOpts const&, std::ostream&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >&) (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0x193152)
#13 0x000055f94e9d3c80 main (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0xeec80)
#14 0x00007faada8b5b97 __libc_start_main /build/glibc-OTsEL5/glibc-2.27/csu/../csu/libc-start.c:344:0
#15 0x000055f94e9e0b9a _start (/media/krasznaa/WD_1TB_Linux/Software/Clang/11.0.0-5788b98/x86_64-ubuntu1804-gcc8/bin/llvm-spirv+0xfbb9a)
llvm-foreach: Segmentation fault (core dumped)
clang-11: error: llvm-spirv command failed with exit code 1 (use -v to see invocation)
[bash][Elrond]:OutOfSource >
```

But if the custom class's functions are implemented inline, the compilation/linking suceeds.

```
[bash][Elrond]:OutOfSource > clang++ -fsycl -DHELPER_IS_INLINE -o linker_error linker_error.cxx 
[bash][Elrond]:OutOfSource >
```

I could reproduce this behaviour with both a custom built version of LLVM, and with Beta5 of oneAPI.

Cheers,
          Attila

cc. @fwyzard, @cleggett

**Comment by AlexeySachkov** [comment]
@krasznaa, this issue should be fixed by KhronosGroup/SPIRV-LLVM-Translator#589: I expect this patch to be merged into intel/llvm with the next pulldown (next week, I guess)

**Comment by AlexeySachkov** [comment]
Should be fixed already. @krasznaa, please let us know if it still doesn't work as expected on your side
